File: src/leetcode/python/contest/LexicographicalNumbers_386.py
class LexicographicalNumbers_386(object):
    def lexicalOrder(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        res = list()
        self.generateNum(res, 1, n);
        return res

    def generateNum(self, res, para, n):
        while para * 10 <= n:
            res.append(para)
            self.generateNum(res, para*10, n)
            para += 1
            if para % 10 == 0: return
        res.append(para)
        para += 1
        while para <= n and para % 10 != 0:
            res.append(para)
            para += 1
        return

File: src/leetcode/java/LexicographicalNumbers_386.java
package leetcode.java;

import java.util.ArrayList;
import java.util.List;

public class LexicographicalNumbers_386 {
  public List<Integer> lexicalOrder(int n) {
    ArrayList<Integer> res = new ArrayList<>(n);
    generateNum(res, 1, n);
    return res;
  }

  private void generateNum(List<Integer> res, int i, int n) {
    while (i * 10 <= n) {
      res.add(i);
      generateNum(res, i * 10, n);
      i++;
      if (i % 10 == 0) return;
    }
    res.add(i);
    while (++i <= n && i % 10 != 0) res.add(i);
    return;
  }
}
